analyses/comparative-RNASeq-analysis/02-thresholds-and-outliers.py

```diff
@@ -0,0 +1,231 @@
+
+Therefore, we modify the Tukey method by, for each gene, dropping the single highest
+expression value (for simplicity) and calculating the high threshold; and similarly
+dropping the single lowest expression value and calculating the low threshold.
+
+https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/229
+"""
+
+import os
+import math
+import sys
+import argparse
+import pandas as pd
+import numpy as np
+import utils
+
+def normalize_expression(input_path, scratch_dir, prefix="", verbose=False):
+    """Convert samples to log2(TPM) + 1 and save in scratch.
+    Note: This is duplicate work from step 01 because I forgot in that step
+    to write results to scratch file. Future update will merge these. (TODO)"""
+    print_v = print if verbose else lambda *a, **k: None
+    print_v("Loading samples {}".format(input_path))
+    raw_samples = utils.read_rds(input_path)
+    samples = np.log2(raw_samples+1)
+    output_path = os.path.join(scratch_dir, "{}log2-normalized.rds".format(prefix))
+    print_v("Writing normalized samples to {}".format(output_path))
+    utils.write_rds(samples, output_path)
+    return samples
+
+def tukey_thresholds(expression_values, iqr_multiplier):
+    """Generate a pair of outlier thresholds (high, low) from
+    expression_values (series) and an iqr multiplier using standard Tukey method."""
+    quartile1 = np.percentile(expression_values, 25)
+    quartile3 = np.percentile(expression_values, 75)
+    iqr = quartile3 - quartile1
+    outlier_range = iqr_multiplier * iqr
+    return (quartile3 + outlier_range, quartile1 - outlier_range)
+
+def modified_tukey_thresholds(expression_values, iqr_multiplier):
+    """Use a modified Tukey method to generate outlier thresholds (high, low)
+    for a single gene. Drops the highest value when calculating the high threshold,
+    and the lowest value for low threshold, to match thresholds for outlier
+    status when a sample is not present in the original dataset."""
+    drop_highest = expression_values.drop(expression_values.idxmax())
+    high = tukey_thresholds(drop_highest, iqr_multiplier)[0]
+    drop_lowest = expression_values.drop(expression_values.idxmin())
+    low = tukey_thresholds(drop_lowest, iqr_multiplier)[1]
+    return( pd.Series([high, low], index=["high", "low"]))
+
+def generate_all_thresholds(normalized_samples,
+                        scratch_dir,
+                        iqr_multiplier=1.5,
+                        prefix="",
+                        verbose=False):
+    """Generate high and low outlier thresholds for each gene
+    using the modified tukey method from a matrix of
+    log2(TMP+1) normalized sample expression"""
+    print_v = print if verbose else lambda *a, **k: None
+
+    print_v("Calculating outlier thresholds.")
+    thresholds = normalized_samples.apply(modified_tukey_thresholds, args=[iqr_multiplier], axis=1)
+    thresholds_path = os.path.join(scratch_dir, "{}threshold-expression-values.rds".format(prefix))
```

Comment:
You're right, the thresholds file was for debugging so I've removed it.
Feather in general looks really promising --as part of my next pull I'll look into switching to it but I'd prefer not to deal with the extra dependencies at this exact moment.

Reply:
That’s fine; the switch to tsv for the main output file was more important. I don't *think* it would actually add a dependency, as pandas supports it, but the advantages are not important here.